Simplify instructions for using ILC (#70811)
dotnet/runtime: src/coreclr/nativeaot/docs/compiling.md

# Compiling with Native AOT

> Please consult [documentation](https://github.com/dotnet/docs/blob/main/docs/core/deploying/native-aot.md) for instructions how to compile amd publish application.

Reviewer: ```suggestion
Please consult [documentation](https://docs.microsoft.com/dotnet/core/deploying/native-aot) for instructions how to compile and publish application.

The rest of this document covers advanced topics only.
```

# Compiling with Native AOT

> Please consult [documentation](https://docs.microsoft.com/dotnet/core/deploying/native-aot) for instructions how to compile and publish application.

Reviewer: Could you please also delete https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/docs/prerequisites.md ? It is no longer referenced now that this is deleted.

The rest of this document covers advanced topics only.

> ## Using daily builds

Reviewer 2: This section is essentially duplicating https://github.com/dotnet/installer#installers-and-binaries, which covers all kinds of `PublishXxx` scenarios (including `PublishAot`).

Reviewer: I think this small duplication is ok to make the steps easier to follow. 

It is duplicated in https://github.com/dotnet/runtime/blob/main/docs/project/dogfooding.md#obtaining-nightly-builds-of-nuget-packages as well.

# Compiling with Native AOT

Please consult [documentation](https://github.com/dotnet/docs/blob/main/docs/core/deploying/native-aot.md) for instructions how to compile amd publish application.

> If you are starting a new project, you may find the [HelloWorld sample](https://github.com/dotnet/samples/tree/main/core/nativeaot/HelloWorld/README.md) directions useful.
> 

Reviewer: ```suggestion
```

Author: I though it was valuable :) but I understand that main docs should be entry point for getting started right now.

> 

Reviewer: ```suggestion
```

This document explains how to compile and publish your project using Native AOT toolchain. First, please _ensure that [pre-requisites](prerequisites.md) are installed_. If you are starting a new project, you may find the [HelloWorld sample](https://github.com/dotnet/samples/tree/main/core/nativeaot/HelloWorld/README.md) directions useful.

> Please consult [documentation](https://github.com/dotnet/docs/blob/main/docs/core/deploying/native-aot.md) for instructions how to compile application.

Reviewer: ```suggestion
Please consult [documentation](https://github.com/dotnet/docs/blob/main/docs/core/deploying/native-aot.md) for instructions how to compile application.

The instructions remaining in this document discuss advanced topics only.
```

Reviewer 3: Should we also include a link to creating the [rd.xml](https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/docs/rd-xml-format.md) file here?

@agocke was thinking ways to unblock developers when they run into issues with native aot and not wait for the SDK to pick the fix. Perhaps, this doc can serve that purpose as well?

Reviewer: We have several other documents on troubleshooting, including rd.xml. https://github.com/dotnet/runtime/tree/main/src/coreclr/nativeaot/docs

## Add ILCompiler package reference

> To use Native AOT with your project, you need to add a reference to the ILCompiler NuGet package containing the Native AOT compiler and runtime.

Reviewer 2: Starting from preview 5, you don't need to do this manually. Just add `<PublishAot>true</PublishAot>` in PropertyGroup, and SDK will pull the required packs (similar to `PublishReadyToRun` or `PublishSingleFile`, we don't add their packages manually).

  </ItemGroup>
```

## Using daily builds

> For using daily builds, you need to make sure the `nuget.config` file for your project contains the following package sources under the `<packageSources>` element:

Reviewer: Adding a nuget feed alone is not sufficient to give you daily builds with the SDK integration.

It needs to something like https://github.com/dotnet/runtime/blob/main/docs/project/dogfooding.md#install-prerequisites . Maybe just refer to that doc.

Author: Based on my experiments to use latest ILC with Preview 5 for example, I have to add following custom target.
```
<Target Name="ForceNightlyILC" BeforeTargets="ProcessFrameworkReferences">
  <ItemGroup>
    <KnownILCompilerPack Update="Microsoft.DotNet.ILCompiler">
      <TargetFramework>net7.0</TargetFramework>
      <ILCompilerPackNamePattern>runtime.**RID**.Microsoft.DotNet.ILCompiler</ILCompilerPackNamePattern>
      <ILCompilerPackVersion>7.0.0-preview.6.22325.1</ILCompilerPackVersion>
      <ILCompilerRuntimeIdentifiers>linux-musl-x64;linux-x64;win-x64;linux-arm;win-arm64;linux-arm64</ILCompilerRuntimeIdentifiers>
    </KnownILCompilerPack>
  </ItemGroup>
</Target>
```

Simply changing `RuntimeFrameworkVersion` was not working as planned.

Reviewer: Hmm, changing `RuntimeFrameworkVersion` is not sufficient to publish regular self-contained apps either. The steps in https://github.com/dotnet/runtime/blob/main/docs/project/dogfooding.md#option-2-self-contained do not work.

The problem is that changing `RuntimeFrameworkVersion` also sets the version for Microsoft.AspNetCore.App.Runtime and Microsoft.WindowsDesktop.App.Runtime. The version numbers of nightly builds of all runtime packs rarely match.

I guess we should keep recommending to use the classic package reference (`<PackageReference Include="Microsoft.DotNet.ILCompiler" Version="7.0.0-*" />`) to get the nightly builds.
